### ISSUE TITLE
Would you like flash attention?

### ISSUE THREAD

**Issue by Birch-san** [lucidrains/muse-maskgit-pytorch#35]
I implemented support for flash attention via xformers:  
https://github.com/Sygil-Dev/muse-maskgit-pytorch/compare/main...Birch-san:muse-maskgit-pytorch:sdp-attn

It gets the same result, confirmed via `allclose()` (I had to make the absolute tolerance a bit more forgiving, to pass that check, but it's still a very small tolerance).

I also implemented support for [`torch.nn.functional.scaled_dot_product_attention`](https://pytorch.org/docs/2.0/generated/torch.nn.functional.scaled_dot_product_attention.html). it'll use flash attention when `mask is None`, but I guess your mask is usually defined.  
even without flash attention: `scaled_dot_product_attention` should still be faster than the `einsum() * scale` approach, because (IIRC) its math fallback is based on `baddbmm`, which fuses the scale factor into the matmul.  
in stable-diffusion inference for example: we measured end-to-end image generation via `baddbmm` to be [~11% faster](https://github.com/huggingface/diffusers/pull/1203)  than `einsum() * scale` on CUDA (and 18% faster on Mac).

@lucidrains is this a contribution you would be interested in receiving as a PR?

**Comment by lucidrains** [comment]
@Birch-san omg, yes! thank you Birch-san!

**Comment by lucidrains** [comment]
@Birch-san make sure to give Tri Dao another cite too! god knows he isn't being cited enough for this impactful work

**Comment by lucidrains** [comment]
@Birch-san hey, decided to take care of it this morning, happy training! (just set `flash = True` on the transformer)

**Comment by Birch-san** [comment]
@lucidrains  
hmm does this actually run with flash attention, or does it fallback to math mode? I thought none of the torch sdp kernels supported masks.

In my diff, where I demonstrated it working with torch sdp: I tried disabling math fallback and it said there was no eligible kernel.  
That’s why I also demonstrated how to do it via xformers, which supports masks.

**Comment by lucidrains** [comment]
@Birch-san yea it should, as they actually just ended up use Tri's flash attention (with the correct hardware), which supports masking

**Comment by lucidrains** [comment]
https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html does 'math' mode not support attention masks? i see `attn_mask` in the argument, so that would be strange

**Comment by Birch-san** [comment]
math mode supports masks. but it's not flash attention. it's just a baddbmm followed by a softmax and a bmm. no memory efficiency, no IO awareness. that's why it's supported on non-CUDA platforms.

if you disable math sdp attn with the context manager:

<img width="623" alt="image" src="https://github.com/lucidrains/muse-maskgit-pytorch/assets/6141784/7513c641-9019-48a2-b0eb-1cbecfa8d4f4">

you get this:

<img width="826" alt="image" src="https://github.com/lucidrains/muse-maskgit-pytorch/assets/6141784/4d86f47a-98d5-49de-b7a4-de12bd956f93">

that's why I showed how to implement flash attention via xformers instead: to get flash attention with support for masking.

**Comment by lucidrains** [comment]
ah, if `math` is not following the tiled memory efficient algorithm, that is news to me, and maybe i should fall back to the naive implementation [here](https://github.com/lucidrains/memory-efficient-attention-pytorch)

xformers uses the 'mem_efficient_attention` which is basically a variant of flash attention. but pytorch 2.0 should detect and use it with the right hardware?

![Screenshot from 2023-07-15 11-21-11](https://github.com/lucidrains/muse-maskgit-pytorch/assets/108653/2e9214e3-0ea2-4511-960b-07277e1e4e11)

anyways, this is all more confusing than it should be. perhaps a meta-library is in order that can optimally select between all these libraries (including triton's impl)

**Comment by Birch-san** [comment]
you can see the implementation of math attention in pytorch's [`attention.cpp`](https://github.com/pytorch/pytorch/blob/efdabbff06611078de275d085f75f724a8059740/aten/src/ATen/native/transformers/attention.cpp#L599-L619).  
it actually looks like it doesn't even use `baddbmm` optimizations (fusing the scale factor and the mask bias), and there's no explicit attempt to describe it as a 3D batched matmul. I wonder if a naïve implementation could actually go faster.

I think there must be some subtlety other than hardware, because for me: xformers 0.0.20 supports masks but we can see torch sdp didn't. perhaps xformers hasn't upstreamed all their capabilities into pytorch sdp?

I also did a naïve implementation of memory-efficient attention, with `baddbmm` optimizations:  
https://github.com/huggingface/diffusers/issues/1892

**Comment by lucidrains** [comment]
@Birch-san yea, makes sense, decided to fallback to my own naive implementation

if they don't have masking support for mem-efficient version, they should def upstream xformers version. masking is very basic.
